llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp

```diff
@@ -1428,6 +1428,46 @@ static Instruction *foldBitOrderCrossLogicOp(Value *V,
   return nullptr;
 }
 
+/// Fold an unsigned minimum of trailing or leading zero bits counts:
+///   umin(cttz(CtOp, ZeroUndef), ConstOp) --> cttz(CtOp | (1 << ConstOp))
+///   umin(ctlz(CtOp, ZeroUndef), ConstOp) --> ctlz(CtOp | ((1 << (bitwidth-1))
+///                                              >> ConstOp))
+template <Intrinsic::ID IntrID>
+static Instruction *
+foldMinimumOverTrailingOrLeadingZeroCount(Instruction *OrigInst, Value *I0,
+                                          Value *I1, const DataLayout &DL,
+                                          InstCombiner::BuilderTy &Builder) {
+  static_assert(IntrID == Intrinsic::cttz || IntrID == Intrinsic::ctlz,
```

Comment:
nit: Would prefer IntrId to be an argument, let the inliner decide if its worth duplicating.

Reply:
To make it work with `m_Intrinsic` and its template parameter, we would have to resort to constructions like
```cpp
if (match(I0, m_OneUse(IntrID == Intrinsic::cttz
    ? m_Intrinsic<Intrinsic::cttz>(m_Value(X), m_Value(Z))
    : m_Intrinsic<Intrinsic::ctlz>(m_Value(X), m_Value(Z))))) {
```
Or am I missing some easier solution?

Reply:
```
if (I0->hasOneUse()) {
  if (auto *II = dyn_cast<IntrinsicInst>(I0); II && II->getIntrinsicID() == IntrID) {
     X = II->getArgOperand(0);
     Z = II->getArgOperand(1);
  }
}
```

Reply:
Of course, if it's okay to not use the matchers, let's do it this way. I've adapted the implementation accordingly.